src/main/java/data/TaskList.java

```diff
@@ -0,0 +1,149 @@
+     * Appends the given {@link Task} to the list of stored tasks.
+     * 
+     * @param task The task to be stored in the list.
+     */
+    public void add(Task task) {
+        this.tasks.add(task);
+    }
+
+    /**
+     * Marks/unmarks the {@link Task} at the given taskIndex
+     * depending on isDone.
+     * 
+     * @param taskIndex The index of the task stored.
+     * @param isDone Whether the task is being marked as done or not done.
+     * @return The task that was marked.
+     * @throws DukeException Thrown when the taskIndex given
+     *                       is negative or out of bounds.
+     */
+    public Task mark(int taskIndex, boolean isDone) throws DukeException {
+        // User tries to mark/unmark a task that is out of bounds.
+        if (taskIndex < 0 || taskIndex >= getSize()) {
+            throw new DukeException(String.format(
+                "Unable to %s task %d :( You have %d task(s) stored.",
+                isDone ? "mark" : "unmark", taskIndex, getSize()
+            ));
+        }
+
+        Task task = getTask(taskIndex);
+        if (isDone) {
+            task.mark();
+        } else {
+            task.unmark();
+        }
+        return task;
+    }
+
+    /**
+     * Deletes the {@link Task} at the given taskIndex.
+     * 
+     * @param taskIndex The index of the task stored.
+     * @return The task that was deleted.
+     * @throws DukeException Thrown when the taskIndex given
+     *      *                is negative or out of bounds.
+     */
+    public Task delete(int taskIndex) throws DukeException {
+        if (taskIndex < 0 || taskIndex >= getSize()) {
+            throw new DukeException(String.format(
+                "Unable to "
+                        + Ui.cTxt("delete", Ui.Color.PURPLE)
+                        + " task %d :( You have %d task(s) stored.",
+                taskIndex + 1, getSize()
+            ));
+        }
+
+        Task removedTask = this.tasks.remove(taskIndex);
+        return removedTask;
+    }
+
+    /**
+     * Returns the string representation of the TaskList.
+     * Contains string representations of all {@link Task} stored
+     * which is suitable for {@link Storage} to write to memory.
+     *
+     * @param keyword The keyword to be used to find tasks
+     *                containing it.
+     * @return An ArrayList of {@link Task} whose description
+     *         contains the keyword given.
+     */
+    public ArrayList<Task> findTasks(String keyword) {
+        ArrayList<Task> result = new ArrayList<>();
+        for (Task t : tasks) {
+            int index = t.findKeyword(keyword);
+            if (index >= 0)
+                result.add(t);
+        }
+        return result;
+    }
+
+    @Override
+    public String toString() {
+        String taskListString = "";
+        for (Task task : this.tasks) {
+            taskListString += task.toFileFormatString() + "\n";
+        }
+        return taskListString.strip();
+    }
+}
```

Comment:
In the toString method, consider using a StringBuilder for concatenation in loops for better performance.